Moving event handling for external subscribers to use go-events

nishanttotla opened the pull request:
Fix https://github.com/docker/swarm/issues/2718

This PR starts to move event handling for external subscribers (API users) to use the battle-tested github.com/docker/go-events repo instead of `cluster.EventHandlers` defined internally in Swarm. To be more precise, it uses the `Queue` abstraction defined in the `watch` package in github.com/docker/swarmkit.

For now, event listening by the watchdog (for rescheduling) continues to be done by the old code.

nishanttotla commented:
@aaronlehmann @alexmavr this PR is a rough pattern of the changes. The relevant commit is https://github.com/docker/swarm/pull/2747/commits/d0f5ce35070d4c6eaf68130bb4689cfd0ad0f8c5 (the others are vendoring). There's a race condition somewhere that I'm trying to debug now, but let me know if this is a reasonable way to use the queue.

I have let the old event handler stay, because the rescheduling code and Mesos integration uses it, but will work on removing it later.

Still remaining:
1. Closing the queue. Thinking about the right place to do it (it should be someplace where the API server on the primary manager is shut down)
2. Handling `--until`. I still have to look up how to use that with the Queue.

aaronlehmann commented on api/handlers.go:
```diff
@@ -890,8 +890,28 @@ func getEvents(c *context, w http.ResponseWriter, r *http.Request) {
+						f.Flush()
+					}
+					if err != nil {
+						log.Debugf("failed to write event to output stream %s", err.Error())
```
If there was an error writing to the socket, this code should return from the function. Currently this loop will keep receiving events and trying to write them forever.

aaronlehmann commented:
This looks like the right approach to me.

nishanttotla commented:
@aaronlehmann @alexmavr I've updated the PR, and fixed all outstanding issues, barring one (getting an accurate counter for the number of listeners, which at this point will return a lower estimate).

For reference and help with reviewing, the main event handler code is in `handlers.go`: https://github.com/docker/swarm/pull/2747/files#diff-379151c15066f365334797e959e6ea94R893

The watch queue itself and related functions are part of `cluster.go`:
https://github.com/docker/swarm/pull/2747/files#diff-423777b5183d89a8b517694ebcebfd6a

The rest of it should be hopefully straightforward. Also ping @dhiltgen @wsong @aluzzardi

aaronlehmann commented on api/events.go:
```diff
@@ -80,13 +80,15 @@ func (eh *eventsHandler) cleanupHandler(remoteAddr string) {
 	e.From = e.From + " node:" + e.Engine.Name
 
+	// Lock access to the attributes map
+	e.Lock()
```
Shouldn't this operate on a copy of `e`, so that the event itself (which is shared between all the subscribers) is not modified by each subscriber?

nishanttotla replied:
@aaronlehmann yeah, I intended to do that, but must have forgotten. I think doing a deepcopy of the event is the way to go.

aaronlehmann commented on api/handlers.go:
```diff
@@ -890,8 +890,47 @@ func getEvents(c *context, w http.ResponseWriter, r *http.Request) {
+	defer cancelFunc()
+
+	// create timer for --until
+	timer := time.NewTimer(0)
```
We shouldn't create the timer in the case where `--until` is not specified. It would fire immediately, and the `select` would always be able to take the `<-timer.C` branch, hogging CPU.

How about something like this:

```go
var (
        timer *time.Timer
        timerCh <-chan time.Time
)
if until > 0 {
        dur := time.Unix(until, 0).Sub(time.Now())
        timer = time.NewTimer(dur)
        timerCh = timer.C
}
```

Then have a `case <-timerCh:` in the select below.

nishanttotla replied:
This is a better idea.

aaronlehmann commented on api/handlers.go:
```diff
@@ -890,8 +890,47 @@ func getEvents(c *context, w http.ResponseWriter, r *http.Request) {
+			if e, ok := eChan.(*cluster.Event); ok {
+				if data, err := normalizeEvent(e); err != nil {
+					return
+				} else {
```
Don't need an `else` because of the `return` above.

nishanttotla commented:
@aaronlehmann addressed your comments.

aaronlehmann commented on api/events.go:
```diff
@@ -80,29 +80,34 @@ func (eh *eventsHandler) cleanupHandler(remoteAddr string) {
-	data, err := json.Marshal(e)
+	data, err := json.Marshal(&e)
+
 	e.From = from
```
This line seems to be a noop. It happens after the marshalling, so it doesn't affect the data returned by this function.

Since the old code used to mutate the event, let's make sure nothing depends on this being set in the `Event` structure.

aaronlehmann replied:
Oh, I guess this was actually to restore the `From` field to the original contents, so it looks like it's not needed anymore, since we're acting on a copy.

nishanttotla replied:
You have a point, but this code is weird. I suspect that this line was needed to reset the changes done by
```
	from := e.From
	e.From = e.From + " node:" + e.Engine.Name
```
above. Since we're not mutating the event anymore, it shouldn't be required to set back to the old value anymore. I'll remove this.

aaronlehmann commented on api/events.go:
```diff
@@ -115,6 +120,17 @@ func (eh *eventsHandler) Handle(e *cluster.Event) error {
 	data = data[:len(data)-1]
 	data = append(data, []byte(node)...)
```
Unrelated to the changes in this PR, but this is really weird. Not sure why `node` can't just be a struct field that gets serialized the normal way.

nishanttotla replied:
I'm not sure about this myself, but my guess is that this is related to backward compatibility. It's been around long enough that it should be removable now though. I can address that later.

aaronlehmann commented:
LGTM (not a maintainer)

aaronlehmann commented:
Note that this should probably use the timeout and queue size limit features introduced in https://github.com/docker/swarmkit/pull/2285 when that PR is ready. That would protect us against introducing out-of-memory regressions in situations where listeners are too slow to consume events, and those events would pile up.

stevvooe commented on api/events.go:
```diff
@@ -80,29 +80,31 @@ func (eh *eventsHandler) cleanupHandler(remoteAddr string) {
+// and all the right fields filled up
+func normalizeEvent(receivedEvent *cluster.Event) ([]byte, error) {
+	// make a local copy of the event
+	e := *receivedEvent
```
Are there any slices or maps here?

nishanttotla replied:
@stevvooe yes, there's only one map that's written to. For that, I've done a deepcopy in the lines below to prevent a race condition.

stevvooe replied:
While it is slightly wasteful if you're not writing to it, it might be worthwhile to just copy the whole structure here to avoid future bugs. These can be nasty in practice.

stevvooe commented on api/handlers.go:
```diff
@@ -890,8 +890,47 @@ func getEvents(c *context, w http.ResponseWriter, r *http.Request) {
+		select {
+		case eChan := <-eventsChan:
+			if e, ok := eChan.(*cluster.Event); ok {
+				data, err := normalizeEvent(e)
```
Move the "meat" outside the if and just break on getting the wrong object.

stevvooe commented on api/handlers.go:
```diff
@@ -890,8 +890,47 @@ func getEvents(c *context, w http.ResponseWriter, r *http.Request) {
+
+	for {
+		select {
+		case eChan := <-eventsChan:
```
Is `eChan` not an event?

stevvooe replied:
@aaronlehmann Does this need "closed" detection? For example, will the watch ever shutdown this channel?

aaronlehmann replied:
Currently it will not, but https://github.com/docker/swarmkit/pull/2285 adds this as an option. So it probably makes sense to add the check here in anticipation of using that later.

nishanttotla replied:
`eChan` has an interface type, which we cast to `*cluster.Event` below.

stevvooe commented on api/handlers.go:
```diff
@@ -890,8 +890,47 @@ func getEvents(c *context, w http.ResponseWriter, r *http.Request) {
+				if err != nil {
+					return
+				}
+				_, err = fmt.Fprint(w, string(data))
```
Why not write directly here, rather than reallocate the data as a string? `w.Write(data)` should be sufficient.

stevvooe commented on api/handlers.go:
```diff
@@ -890,8 +890,47 @@ func getEvents(c *context, w http.ResponseWriter, r *http.Request) {
+	}
+	var closeNotify <-chan bool
+	if closeNotifier, ok := w.(http.CloseNotifier); ok {
+		closeNotify = closeNotifier.CloseNotify()
```
Depending on go version, you can now use `r.Context().Done()` to listen to this cancellation.

nishanttotla replied:
I'm not sure the request context is set correctly in all cases in Swarm, some of the code is from before contexts were introduced in Go. I'll check.

stevvooe replied:
This should be from `net/http`, so should be wired to the `CloseNotifier`.

stevvooe commented on cluster/mesos/cluster.go:
```diff
@@ -168,12 +170,20 @@ func NewCluster(scheduler *scheduler.Scheduler, TLSConfig *tls.Config, master st
 // Handle callbacks for the events
 func (c *Cluster) Handle(e *cluster.Event) error {
+	// publish event to the watchQueue for external subscribers
+	c.watchQueue.Publish(e)
```
Why don't we adapt this to the eventHandlers list? I am not sure that making this dependent on the watch queue makes sense.

nishanttotla replied:
@stevvooe I agree, but I also think that the eventual queue we will used will be moved upstream to go-events. This is supposed to be a temporary fix.

`eventHandlers' will go away actually, and all events should be handled using the watchQueue. For now though, some internal rescheduling functionality uses events which I did not want to mess with in this PR, that's why we have this case.

stevvooe replied:
@nishanttotla That still doesn't explain why this isn't just a handler. You can do this with a closure to adapt `Publish` to `Handler`.

nishanttotla replied:
@stevvooe not sure I follow exactly here. What do you envision this handler to look like?

stevvooe replied:
Not sure what you mean: just make an event handler. I am not sure how this got merged in its current state. It just doesn't make any sense to have both a handler set and a publish set. I'd recommend cleaning this up before it gets too far. It is going to be a pain to remove later and will confuse future contributors.

nishanttotla replied:
I will clean that up soon. The handler will entirely go away as it is, and only the publish set will remain.

alexmavr commented on api/primary.go:
```diff
@@ -122,16 +124,22 @@ func NewPrimary(cluster cluster.Cluster, tlsConfig *tls.Config, status StatusHan
+	// eventsQueue uses the watch package from SwarmKit, which is based on
+	// the go-events package. See https://github.com/docker/swarm/issues/2718
+	// for context
+	eventsQueue := watch.NewQueue()
```
Now that https://github.com/docker/swarmkit/pull/2285 is merged, let's revendor swarmkit and change this constructor to:
```
NewQueue(WithTimeout(10 * time.Second), WithLimit(10000), WithCloseOutChan())
```

We probably want to define these values as config vars and plumb them through the standard config model used by classic swarm. 
Assuming 1KB per event (gross overestimation), the size limit of `10000` for the queue should occupy no more than 10MB per listener in memory. For the edge case of 100 simultaneous inefficient listeners, this would lead up to to an additional 1 GB of memory

nishanttotla replied:
@alexmavr done. PTAL.

alexmavr commented on api/handlers.go:
```diff
@@ -890,8 +890,49 @@ func getEvents(c *context, w http.ResponseWriter, r *http.Request) {
+			if !ok {
+				break
+			}
+			data, err := normalizeEvent(e)
```
Should we consider event filtering here?

nishanttotla replied:
@alexmavr maybe we could do it as a follow up?

nishanttotla commented:
@alexmavr added a counter to track number of active listeners, PTAL.

aluzzardi commented:
/cc @wsong for review

nishanttotla commented:
Merging this now, since it's been reviewed and it worked in testing, both manual and `docker-e2e`. Any downstream issues can be fixed in a follow-up.

stevvooe commented:
@nishanttotla I'd recommend taking time to cleanup the double handler logic. That will result in a bug in the future if one adds some to the task queue and the handler set.
